ml-agents-envs/mlagents_envs/registry/binary_utils.py

```diff
@@ -0,0 +1,198 @@
+    if os.path.exists(binary_path):
+        shutil.rmtree(binary_path)
+
+    # Download zip
+    try:
+        request = urllib.request.urlopen(url)
+    except urllib.error.HTTPError as e:  # type: ignore
+        e.msg += " " + url
+        raise
+    zip_size = int(request.headers["content-length"])
+    zip_file_path = os.path.join(zip_dir, str(uuid.uuid4()) + ".zip")
+    with open(zip_file_path, "wb") as zip_file:
+        downloaded = 0
+        while True:
+            buffer = request.read(BLOCK_SIZE)
+            if not buffer:
+                # There is nothing more to read
+                break
+            downloaded += len(buffer)
+            zip_file.write(buffer)
+            downloaded_percent = downloaded / zip_size * 100
+            print_progress(f"  Downloading {name}", downloaded_percent)
+        print("")
+
+    # Extraction
+    with ZipFileWithProgress(zip_file_path, "r") as zip_ref:
+        zip_ref.extract_zip(f"  Extracting  {name}", binary_path)  # type: ignore
+    print("")
+
+    # Clean up zip
+    print_progress(f"  Cleaning up {name}", 0)
+    os.remove(zip_file_path)
+
+    # Give permission
+    for f in glob.glob(binary_path + "/**/*", recursive=True):
+        # 16877 is octal 40755, which denotes a directory with permissions 755
+        os.chmod(f, 16877)
+    print_progress(f"  Cleaning up {name}", 100)
+    print("")
+
+
+def print_progress(prefix: str, percent: float) -> None:
+    """
+    Displays a single progress bar in the terminal with value percent.
+    :param prefix: The string that will precede the progress bar.
+    :param percent: The percent progression of the bar (min is 0, max is 100)
+    """
+    BAR_LEN = 20
+    percent = min(100, max(0, percent))
+    bar_progress = min(int(percent / 100 * BAR_LEN), BAR_LEN)
+    bar = "|" + "\u2588" * bar_progress + " " * (BAR_LEN - bar_progress) + "|"
+    str_percent = "%3.0f%%" % percent
+    print(f"{prefix} : {bar} {str_percent} \r", end="", flush=True)
+
+
+def load_remote_manifest(url: str) -> Dict[str, Any]:
+    """
+    Converts a remote yaml file into a Python dictionary
+    """
+    tmp_dir, _ = get_tmp_dir()
+    try:
+        request = urllib.request.urlopen(url)
+    except urllib.error.HTTPError as e:  # type: ignore
+        e.msg += " " + url
+        raise
+    manifest_path = os.path.join(tmp_dir, str(uuid.uuid4()) + ".yaml")
+    with open(manifest_path, "wb") as manifest:
+        while True:
+            buffer = request.read(BLOCK_SIZE)
+            if not buffer:
+                # There is nothing more to read
+                break
+            manifest.write(buffer)
+    try:
+        result = load_local_manifest(manifest_path)
+    finally:
+        os.remove(manifest_path)
+    return result
+
+
+def load_local_manifest(path: str) -> Dict[str, Any]:
+    """
+    Converts a local yaml file into a Python dictionary
+    """
+    with open(path) as data_file:
+        return yaml.safe_load(data_file)
```

Comment:
might want to handle an exception here in case yaml.safe_load throws an error.

Reply:
What is wrong with raising that error? I don't want to suppress it.

Reply:
Firstly there is nothing wrong in what you did. Secondly, raising an exception allows you to give a more informative message to the user instead of a generic error message that comes from python. I think it's more of a matter of coding style rather than correctness. 

It was just a suggestion to begin with :)